Source/WebCore/Modules/mediasource/SourceBuffer.cpp

```diff
@@ -664,257 +680,260 @@ void SourceBuffer::setActive(bool active)
 
 Ref<MediaPromise> SourceBuffer::sourceBufferPrivateDidReceiveInitializationSegment(InitializationSegment&& segment)
 {
-    if (isRemoved())
-        return MediaPromise::createAndReject(PlatformMediaError::BufferRemoved);
-
-    ALWAYS_LOG(LOGIDENTIFIER);
-
-    // 3.5.8 Initialization Segment Received (ctd)
-    // https://rawgit.com/w3c/media-source/c3ad59c7a370d04430969ba73d18dc9bcde57a33/index.html#sourcebuffer-init-segment-received [Editor's Draft 09 January 2015]
-
-    // 1. Update the duration attribute if it currently equals NaN:
-    if (m_source->duration().isInvalid()) {
-        // ↳ If the initialization segment contains a duration:
-        //   Run the duration change algorithm with new duration set to the duration in the initialization segment.
-        // ↳ Otherwise:
-        //   Run the duration change algorithm with new duration set to positive Infinity.
-        if (segment.duration.isValid() && !segment.duration.isIndefinite())
-            m_source->setDurationInternal(segment.duration);
-        else
-            m_source->setDurationInternal(MediaTime::positiveInfiniteTime());
-    }
+    return invokeAsync(m_dispatcher, [weakThis = ThreadSafeWeakPtr { *this }, this, segment = WTFMove(segment)]() mutable {
+        RefPtr protectedThis = weakThis.get();
+        if (!protectedThis || isRemoved())
```

Comment:
Seems like this is `ensureOnDispatcher()` but returns a Promise. Should there be a `ensureOnDispatcher()` variant that similarly returns a promise?.

Reply:
invokeAsync is a templated utility that returns the same type as what the lambda returns. If we only ever returned the same promise type that would do, but we aren't.

This code will have to be changed in a later PR anyway as invokeAsync requires a guarantee that the invoked lambda gets called, which once we run on a worker will not happen.